src/caching_key_wrapper.rs

```diff
@@ -121,37 +121,50 @@ where
         }
     }
 
-    async fn get_and_increment_cached_encryption_key(&self, bytes: usize) -> Option<DataKey<S>> {
-        let mut cached = self.encryption_cache.lock().await;
-
-        while let Some(mut cached_entry) = cached.pop() {
-            cached_entry.messages_encrypted += 1;
-            cached_entry.bytes_encrypted += bytes;
-
-            if !self.has_exceeded_limits(&cached_entry) {
-                // Cloning here could be bad for zeroing memory + performance.
-                // Maybe change this method to be a "with_data_key_for_bytes" so that
-                // we can borrow the key instead.
-                let key = cached_entry.key.clone();
-
-                // Since the entry is fine to keep, add it back to the stack
-                cached.push(cached_entry);
-
-                return Some(key);
+    async fn get_and_increment_cached_encryption_key(
+        &self,
+        bytes: usize,
+        aad: Option<&str>,
+    ) -> Option<DataKey<S>> {
+        let mut encryption_cache = self.encryption_cache.lock().await;
+
+        let cache_key = aad.unwrap_or_default().as_bytes();
```

Comment:
What if 2 keys have the same AAD?

Reply:
Actually, I can see by your tests that this likely isn't a problem.

Reply:
This can happen if 2 concurrent requests happen when there are no keys in the cached. In which case the last entry get popped off the Vec and served.